### PR TITLE
First pass at adding relator mapping capabilities to author.bsh.

### PR DESCRIPTION
**Pull request by demiankatz**
The current relator handling in VuFind is really designed to work with the use of 3-letter codes in subfield $4. This does not play well with RDA practice, where fully-spelled-out relators are meant to be put in subfield $e. This pull request adds some mapping functionality to allow better support for full relator terms while still storing abbreviations in the VuFind index when possible for uniformity and translation support.

TODO:

- [x] Update SolrMarc -- as implemented with the current version of VuFind, this indexing process is EXTREMELY slow. The caching added at https://github.com/solrmarc/solrmarc/commit/ada6a81957d83fb55cf0389db6b37a21257e7d51 significantly improves the situation.
- [x] Resolve #772 -- there are some overlapping concerns, and resolving that PR may help simplify this one.
- [x] Test deduplication in VuFind(-Frontend)
- [x] Resolve case sensitivity problems; as written, the code in this PR (mirroring the code in master) will store normalized forms of relator terms in the index. However, this may not be desirable in all cases! We may need to do some extra processing to ensure that terms which match the whitelist in author-classification.ini are stored in exactly the same form as they are represented in that file, even if they are matched in a fuzzy manner.
- [x] Determine how to handle unknown terms/codes not found in whitelist. Should we ignore them, include them, or conditionally decide? Is additional configuration needed?

### CODE REVIEW

**Comment by lahmann** [comment]
Demian,

thanks for taking the time working on the author indexing issues! The code-mapping approach looks promising, though after looking at the code and recalling the discussions with my colleagues I think there might be a (German cataloging?) edge case worth thinking about.

The German gender neutral translations for relator codes we contributed are actually relator terms used in $e. The german library associations GBV and SWB (which we in Leipzig are part of) agreed upon using gender neutral forms in $e but requiring relator codes in $4 for data exchange (export) usages. As not all (/custom) German gender neutral relator terms could be mapped to relator codes those were agreed upon being mapped to relator code "oth" in $4. Obviously the current implementation in vufind-org/master is not appropriately covering this case at all - I just recognized the issue after thinking about this pull request - but the relator-mapping might not address this case appropriately as well.

Applying this pull request for the described cases the mapping would result in dropping the specific relator term in $e and substituting it by relator code "oth" which is a clear loss of relevant information.
After looking at some RDA documentation especially about how to handle the relator codes and terms (LINKS) I get the impression that the relator codes in $4 are being kept more for legacy and export/exchange reasons than actually being the authoritative subfield regarding relators.
This complicates everything - going back to the cataloging practise in Germany described above we might have to add cases where the correlation of code and term is unknown:
* index authors in MARC 100 as primary authors with relator terms in $e but either $4 being empty or $4 containing a code that is not mapped to the value of $e regarding the mappings in relatorConfig
* index authors in MARC 700 as secondary authors with relator terms in $e but either $4 being empty or $4 containing a code that is not mapped to the value of $e regarding the mappings in relatorConfig
* index corporate authors with relator terms in $e but either $4 being empty or $4 containing a code that is not mapped to the value of $e regarding the mappings in relatorConfig
This way we will index codes in cases where the correlation is known and relator terms in all other cases. As relator codes are always alphabetical-lowercase we could skip normalizing the relators being indexed and preserve the casing and additional characters in unknown relator terms. Obviously those could not be translated but these by adding a mapping to a (custom) relator code this could be addressed quite easily.

Any thoughts on this?

**Comment by demiankatz** [comment]
Regarding the German edge case: can this problem be solved by simply including all German gender-neutral terms in the whitelist of acceptable relators while excluding "oth"? Or are we concerned that there may be some entries where "oth" is the only option, in which case we would want to keep that? Alternatively, if everything is on the whitelist, you end up indexing the gender neutral term PLUS other, which creates extra useless data but might not be particularly harmful. The only other possibility would be to create two whitelists: "all terms" and "preferred terms" -- "oth" could be on "all terms" while the other relators could be on "preferred terms." The indexer could be adjusted to create two lists of relators -- the "all" list and the "preferred" list. It would return the preferred list if non-empty, otherwise the all list. This would cover more cases but also require even more configuration to get things working.

This also touches a bit on the issue I mentioned in #772 regarding terms that aren't found on any lists. Handling of these cases could overlap with the hypothetical preferred term system in a couple of different ways, which is something else to think about.

Also, speaking of configuration, a side note: have we done testing of non-ASCII characters in author-classification.ini? I notice that some of the German relator codes have special characters in them, and we should probably make sure this works as expected. I'm not sure how well the Java .ini reader we are using handles UTF-8. Definitely something we should add better coverage for in our test records.

Regarding correlation of code and term between $4 and $e, I think our best bet is not to assume anything there. If we just take the union of all $4/$e values and then map/normalize/deduplicate them, I think we'll get the desired results. The only problem may be that in #772 I'm using a Set (rather than a List) in order to take advantage of automatic deduplication, and this may result in relators being presented in an unpredictable order. If that's going to be a problem, we may need to use a List there and do manual deduplication.

As far as case sensitivity, etc., goes, I think the best option may be:

* If mappings can be applied, map the term
* If no mapping is found, but the term exists in the whitelist, convert it to match the exact form it takes in the whitelist
* If no mapping or whitelist entry is found, the term will currently never be displayed -- but if we come up with a mechanism for handling "unexpected" relators, we should retain their case as-is and do no extra normalization (unless we want to decide on some minimal steps like trimming whitespace and/or trailing punctuation).

In any case, I think we're getting closer! Let's figure out what we need to do to wrap up #772 and then hopefully this will become easier.

**Comment by lahmann** [comment]
Demian, as a preparation for further testing with our test cases from #772 I updated the author-classification.ini with German RDA relator terms in the RelatorSynonyms section. Next step is testing this with #772.

**Comment by demiankatz** [comment]
Thanks! I hope to get #771 merged into master and this PR later today, since the SolrMarc progress I was hoping for has been successfully made.

I agree that the German RDA terms will be helpful for testing purposes, but I'm not sure if we should include them in the default .ini file, since that would open the flood gate to including every variation in every language, which I imagine could ultimately end up slowing down the indexing process. I definitely think it is useful to include these as an option somewhere, though... just a question of how to approach it:

  * Commented-out block in .ini file?
  * Alternate .ini file with a different name?
  * Wiki page containing language-specific mappings that users can copy and paste?
  * Something else?

What do you think? Ideally it would be nice to have something that's easy to maintain and easy to use, easily expandable to support additional languages, clearly documented in comments in the file, but unlikely to interfere with indexing performance.

**Comment by demiankatz** [comment]
@lahmann, I have successfully merged master into #772 and #772 into this PR. A few notes:

1.) The new SolrMarc seems to be having trouble with author.bsh, so for the moment, we need to do all of our development in Java rather than BeanShell (at least until I hear back from Bob, who I have asked about the problem). I have ported the existing BeanShell modifications into Java. If we make further changes, we need to remember to make them in both places. I suspect that we may be able to drop BeanShell in VuFind 4.0 because the new Java dynamic compilation is superior. However, to avoid a BC break, I think we should continue to maintain both versions, at least through 3.1.

2.) I'm interested in your thoughts on configurability of inclusion of unknown terms/codes. I think we might want to favor one subfield over another in this situation (human-readable strings are preferable to codes in situations where no translation exists) but preferences seem likely to vary from place to place.

3.) I still need to give some more thought to case sensitivity issues. I'll probably put some more time into that tomorrow. For today, I'm going to switch gears and make progress on a couple of other PRs now.

4.) I think the upgraded code should perform a lot better than it used to. Please let me know if you are still seeing poor performance in this branch.

Thanks!

**Comment by demiankatz** [comment]
Quick update to point 1 above -- the BeanShell/SolrMarc problems have been solved, so we can now test both versions of the indexing logic moving forward.

**Comment by lahmann** [comment]
Demian, thanks for your work - SolrMarc3 has brought great perfomance improvement: indexing the test records in author_relators.mrc with SolrMarc2 took 12s yesterday whereas SolrMarc3 was finished after 0.7s today. So this is a huge leap forward!

Now on to your comments:

2.) I think we still have a need to define what we mean by unknown relator terms/codes. The whitelists (firstAuthorRoles, secondAuthorRoles, nonCreativeRoles) currently decide whether a relator term/code is known or not - in marc.properties is defined which whitelist(s) will be used for a list of Marc fields being indexed in a specific Solr field. When we are speaking of _unknown relator terms/codes_ in my understanding we are speaking about relator terms/codes that are not listed in any whitelist. Since this pull request we also have the RelatorSynonyms which could also be added to the knowledge base the whitelists represent.

2.a) Probably we have to have a method `getKnownRelators` which will return any relator either listed in the whitelists or listed as a RelatorSynonym as a known relator.

2.b) Regarding configurability I think a binary switch to allow indexing of unknown relator terms/codes might be useful. The rules should be static in my opinion - to reduce complexity - unknown relator terms in Marc 100 are indexed as primary, in Marc 700 as secondary roles. I agree with you that $e should contain more human readable content than $4 - so maybe indexing should be limited to $e (although this could be left configurable without adding too much complexity).
Additionally a binary switch to log unknown relator terms/codes should be considered as well to help completing the (regionally differing) relator terms/codes in use (logging both - $e and $4).

3.) Indexing the raw form is favorable in my opinion: as you can see in my commit which added the German relator terms to the RelatorSynonyms at least for German cataloging records there is a information loss when only indexing the LOC relator code. A somewhat worst case example is RDA relator terms mapping to the code `ctb`:
```ini
ctb = "Contributor|Mitwirkender|Letterer|Maskenbildner|MaskenbildnerIn|On-Screen-Teilnehmer|On-Screen-TeilnehmerIn|Orchesterleiter|OrchesterleiterIn|Chorleiter|ChorleiterIn|Softwareentwickler|SoftwareentwicklerIn|Special-effects-Provider|Visual-effects-Provider|Verfasser eines Postscriptums|VerfasserIn eines Postscriptums"
```
The mapping is very useful to reduce the size of the whitelists. Looking up each $e and $4 in the whitelists will ensure to index explicitly listed relator terms/codes. Subsequently looking up each $e and $4 with their mapping in RelatorSynonyms will ensure to cover such cases as the `ctb` example above without the need of listing each German relator term explicitly.
If the terms and codes are indexed in their raw form deduplication has to happen with the translated strings in VuFind. This way we will very likely have lots of duplicates being indexed but do not have any information loss and leave it to the frontend's translation configuration to group relator terms/codes (this would also cover the "beiträger" case). Also I assume the raw forms being more suitable for actual searches compared to indexing only relator terms which would always need a mapping from the indexed codes to more speaking terms (e.g. explicitly searching for author, designer etc.).

5.) Regarding the German RDA terms I had the same thought while adding the terms to the ini file :)
I think a commented out block in the ini file would be sufficient. Maybe worth considering is using the relator codes as arrays each array field containing a region specific mapping - e.g.:
```ini
;US Mapping
ctb[] = "Contributor"
;DE Mapping
ctb[] = "Mitwirkender|Letterer|Maskenbildner|MaskenbildnerIn|On-Screen-Teilnehmer|On-Screen-TeilnehmerIn|Orchesterleiter|OrchesterleiterIn|Chorleiter|ChorleiterIn|Softwareentwickler|SoftwareentwicklerIn|Special-effects-Provider|Visual-effects-Provider|Verfasser eines Postscriptums|VerfasserIn eines Postscriptums"
```

In any case this looks very promissing - I am happy to implement the above outlined functionality I just figured that it would be best to discuss the approach in advance than just start implementating.

**Comment by lahmann** [comment]
2.) You are right about the difficulty to decide whether a relator term/code is unknown. But if we understand the RelatorSynonyms as the complete list of known relator codes and terms any of the whitelists would be a subset thus a) any relator term or code found in the whitelist will get indexed, b) any relator term or code not found in the whitelist but in the RelatorSynonyms is not to be indexed, c) any relator term neither found in whitelist nor in RelatorSynonyms will be considered as unknown and the fallback logic for unknown terms/codes would apply.
If the unknown relator fallback logic is configured as disabled any non matching relator will get skipped. Of course we would have to at least define the RelatorSynonyms as authoritative list that needs to exist in author-classification.ini in order to build the knowledge base.

3.) I agree with your argument pro normalization regarding facetting. On the other hand I doubt that German RDA relators can be normalized to relator codes without significant loss of information. As I mentioned earlier some German RDA relator terms are mapped to the same LOC relator code (e.g. `ctb`) as the codes are only used for export (or fallback) anyway. In the specific example of `ctb` we have to index $e's raw content.
But when indexing the raw German relator terms we also would have to check if the current Marc field also contains a subfield with a relator code which is mapped to the to be indexed raw relator term as those two relator data (raw term and code) would definitely not get deduped (there are simply to many multiple terms mapped to one code).
The only solution that comes to my mind is to build the evaluation of the relator subfields more complex and trying to establish links between $e and $4 values as defined in the RelatorSynonyms. In case of an established link we will index $e and drop $4. And one has to decide before indexing whether to add computational complexity and uncertainty for working relator terms/codes translations when indexing raw terms or to accept some information loss during indexing due to normalization.

5.) If we agree upon the direction I am pointing out in 3. it might be justifiable to leave it up to the indexing institution to build one line mappings as currently done in author-classification.ini and spare us the array issue. Most likely the normalization would be default and any attempt to index the raw form require some intellectual work beforehand anyway, so adding some ini editing would not exceed the plus in additional work proportionally.

**Comment by demiankatz** [comment]
2.) Yes, you're right, the synonyms do help with this problem. I wonder if we actually want to slightly repurpose the existing noRelatorAllowed parameter so that instead of just being "if no relators are found in one of these tags, put a blank string in this field" it instead means "if unknown relators are found in one of these tags, pass them through; if no relators are found, add a blank string". Does that make sense?

3.) Is there likely to be a situation where a single author will have multiple relator terms that map to the same thing? If not, one algorithm that might work without requiring us to establish explicit links between $e and $4 could be something like this (in pseudocode):

<pre>
normalizedValues = getNormalizedValues(rawValuesFromSubfieldE, rawValuesFromSubfield4)
returnValues = []
foreach (normalizedValues as value) {
    bestMatch = findClosestMatch(rawValuesFromSubfieldE, value)
    if (!bestMatch) {
        bestMatch = findClosestMatch(rawValuesFromSubfield4, value)
    }
    returnValues[] = bestMatch
}
</pre>

So bottom line: we map and deduplicate values, and then for each of these, we try to pick the "best" raw value, favoring $e over $4. We could add a switch to the functions to skip this step which would easily allow people to index processed rather than raw values if they wanted to for some reason.

The only thing this does not account for is dealing with unknown relator correlation -- but again, I might consider saying "if there are unknown relators in both $e and $4, just take the ones from $e... but if $e is empty, use $4 if you must."

It might also be worth considering whether SolrMarc's native translation map functionality could be helpful in some cases here.

5.) I definitely would like to keep this simple if we can... but I also want to include a default mapping that's at least a useful start. Maybe we should just roll this back to "English mappings by default, with German + English mappings commented out in a block below as an example and for the convenience of German libraries."

**Comment by lahmann** [comment]
2.) I do like the idea of extending the purpose of `noRelatorAllowed` for unknown relators but I have some gut feeling that we should keep those separate as e.g. indexing the person in Marc 100 without having a relator as primary author is a fairly safe bet whereas simultaneously indexing another person in Marc 100 because the relator is unknown should be a different case as the relator does not match an explicit whitelist. In other words it could be quite desireable to be able to distinguish between persons without relators (i.e. the record being not complete with relator information) and persons with unknown relators (i.e. the relator mapping being not complete).
On the other hand the logic to define in `noRelatorAllowed` for which of the given Marc fields this rule applies (i.e. authorFields vs. noRelatorAllowed) makes also sense for unknown relators if we don't want to stick to the static relation Marc 100 - primary, Marc 700 - secondary Marc 1x0/7x0 - corporate.
Keeping the code simple is another pro argument for extending the purpose of `noRelatorAllowed` - when the configuration almost equals programming there is a slight chance that something is over engineered ;)
It looks like it's more sensible to use `noRelatorAllowed` for this purpose - we should also add logging for unknown relators.

3.) Although the likelihood is fairly thin we certainly cannot say for sure that those cases won't happen (especially the earlier mentioned `ctb` example shows that several semantic distinct relator terms are mapped to the same code - so if a person combines all these roles it could be possible to have multiple relator terms for one person being mapped to one single code). However it depends on the implementation of `findClosestMatch(raw,normalized)` - if findClosestMatch becomes a `findAllMatches(raw,normalized)` we will get for raw indexing all raw contents of $e even if the terms are mapped to the same code (and get rid of the unnecessary relator code). If normalized indexing is chosen we will stick with the normalized list which implies reduction and deduplication through normalization anyway. So I think this is a good solution.
Regarding the unknown terms/codes we could get hold of them already when doing `getNormalizedValues(rawE,raw4)` as we are using RelatorSynonyms to normalize the terms and codes. Thus if the normalization fails due to missing mapping we can already take care of these terms/codes by the unknown relator logic.

SolrMarc translation maps are an excellent idea! This way we keep the ini file simple and isolated from the knowledgebase of relator code-term mappings. Does the translation maps parsing logic support delimiters for multiple terms?

5.) I agree.

Unfortunately I won't have time today to actually implement the above mentioned ideas - if you think we have found a consensus on how things should work, I could start with implementation on Monday.

**Comment by demiankatz** [comment]
André,

I had some time to work on this today, so I've made some changes to the Java version of the code. I haven't ported anything to BeanShell yet; might as well finalize the logic before we go to that trouble.

I don't think I've totally captured all of the functionality we've discussed, but I think I've built out significant infrastructure to make the remaining implementation a lot easier... and maybe I'm pretty close to what we want.

For the moment, I have added two new parameters to all of the *FilteredByRelator() methods: the first is a list of fields from which to index unknown relators; the second is "true" to index raw values, "false" to index mapped values. We could very easily add some additional function definitions to set these to reasonable defaults and make them optional. They're explicit for now because I didn't want to waste time defining extra function signatures until we had the functionality we wanted. INFO messages are sent to the logger whenever we index an unknown relator term.

Please review and test the code and let me know what you think! Here's are some specific notes/comments:

1.) I didn't implement the findClosestMatch logic discussed above. We may need something like it, but I opted for a simpler implementation for today due to limited time. Let me know how well it works as-is, and what needs to change if it is not adequate.

2.) In my testing, I definitely saw some duplicate relator terms in the front end. I'm not sure if this is in part due to unicode normalization (or other character encoding) issues, or if it's something less subtle.

3.) The comments on the relator map in author-classification.ini will definitely need to be revised if we go forward with some variation of this functionality, since the behavior now varies based on indexing settings. It might also be smart to put a comment in marc.properties above the author block explaining some of its eccentricities -- this has gotten really complicated.

4.) We should give some thought to what we absolutely need for release 3.1 and what can be deferred -- we only have a few days left before the release, and I want to get something reasonable and sane pushed out.

Anyway, I'm about out of time for now, but hopefully this has helped move things forward. I look forward to talking about it further next week!

**Comment by lahmann** [comment]
Demian, thanks a lot for your work - I think we are almost done! Indexing works as expected: deduplication functions when not indexing raw forms; raw forms index fine but would probably need deduplication in frontend; unknown relator terms/codes index fine too.

The very special case of German relator terms being mapped to the same relator code (example `ctb` from above) are taken care of appropriately when raw indexing is enabled. The differeing relator terms and once the relator code will get indexed. I think this suffices as one can control the indexing with raw indexing disabled by altering the RelatorSynonyms, thus map those relator terms to different codes or remove them from the RelatorSynonyms and index them as unknown as a fallback. Therefore I think we do not need to reconsider 1.)

2.) Fake record 10 is a good test case for deduplication - I have implemented a template based deduplication in `themes/bootstrap3/templates/RecordDriver/SolrDefault/core.phtml` that could easily get moved into a view helper. All translated roles will get deduped per author which should also allow to address optimization of relator display without the need to reindex.
What do you think - is this an appropriate approach?

3.) Yes, also settings for roles and author-fields need to be in synch which is especially counter-intuitive for `indexRawRelators` parameter in `getAuthorsFilteredByRelator()` - otherwise the relators will get indexed raw but do not match the author field by quantity.

4.) I think the current functionality is better than the current master code and does allow even precise customization. So as long as you don't know about any missing important feature I think this should be almost ready to go.

**Comment by demiankatz** [comment]
André,

Thanks for the review! Some notes (new numbering, not corresponding to previous discussions):

1.) I've made some adjustments to your deduplication logic to use more functional rather than procedural techniques for processing relators. This also enabled a significant simplification to the way we output relators in the templates. I've ported this over to jquerymobile as well. It might make sense to move the logic to a view helper, but for now I left it in the template to keep things simple.

2.) I've added some lengthy comments to marc.properties and author-classification.ini in an effort to help clarify everything.

3.) I've commented out the synonym map that includes German terms and restored the English-only version. Unfortunately this means that the test records no longer index correctly by default, but I thought the mix of English and German would confuse the majority of users if that was the default setting. We can continue to talk about a long-term solution for managing multi-lingual mappings, but for now I think this is good enough for release 3.1.

4.) I've added the necessary function signatures to make the newly-added parameters on the indexing routines optional, with reasonable defaults. This prevents the indexing definition from growing unless you need to override the defaults.

5.) I've ported the Java code into BeanShell, so everything should now be in sync.

6.) At this point, the only unresolved checkbox on our TODO list has to do with reducing redundancy in language files. I think that may still be something we can achieve, but I don't think we need to address it right now. In the interest of simplicity, I'm willing to defer that for a later date.

Bottom line: I think we're all ready to go, but I would appreciate if you could give this one more review tomorrow before I push the button and make it official. (I should also probably look at it one more time tomorrow after a good night's rest). Please let me know if you run into any last-minute problems; if not, I'll plan on merging this by end of day tomorrow. Thanks again for all of your help!

**Comment by lahmann** [comment]
Thanks a lot for pushing this over the finish line! I tested indexing with default marc.properties, disabled unknownRelators indexing, enabled rawRelatorsIndexing and tried the RelatorSynonyms with German terms - all works as expected and your refactoring of the dedup code in the core template works fine too and improves code readability.
I pushed some minor comment edits but apart from that I think we are ready for merging.

Regarding revision of the language files I would suggest to address this at a later time not only because the code freeze is getting closer but also to gather some real life experience with the new indexing code and relator handling which might give us some insight how to improve the code and ini files even more.

Thanks again for all your work!

**Comment by demiankatz** [comment]
André,

Glad I could help with this, and thanks for your work as well -- the extra testing is greatly appreciated! I'm going to look this over a little bit more today and will very likely merge it before the end of the day.

Regarding the translation issues, I agree that waiting is wise for all of the reasons you mention. To ensure that we don't forget about it, I've added a note to our JIRA ticket about 4.0-specific language file needs: https://vufind.org/jira/browse/VUFIND-1164 (and have removed the TODO item from this PR, so we can close it out without unresolved issues).

**Comment by demiankatz** [comment]
I expanded one comment and improved my BeanShell port slightly; final results have been merged to master. Thanks again!

**Comment by lahmann** [comment]
Great! Thank you very much! :)
